ci, gha: Add "x86_64: Linux (Debian stable)" GitHub Actions job

hebasto opened the pull request:
Solves one item in https://github.com/bitcoin-core/secp256k1/issues/1392 partially.

hebasto commented:
> What are the ghcr usage limits?

GitHub Container Registry is free for public images.

hebasto commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,110 @@
+  x86_64_debian_stable:
+    name: "x86_64: Linux (Debian stable)"
+    runs-on: ubuntu-latest
+    container: ghcr.io/${{ github.repository_owner }}/secp256k1-ci-image:latest
```
> how will this work, when a pull modifies the `latest` tag and breaks all other build, because they skip the `docker_image` step?

Do you mean concurrent builds?

hebasto replied:
I see. Going to address this issue shortly.

hebasto replied:
Thanks! [Fixed](https://github.com/bitcoin-core/secp256k1/pull/1396#issuecomment-1668502527).

real-or-random commented on ci/cirrus.sh:
```diff
@@ -4,6 +4,10 @@ set -eux
 
 export LC_ALL=C
 
+if [ "${GITHUB_ACTIONS:-undefined}" = "true" ]; then
+    git config --global --add safe.directory "$GITHUB_WORKSPACE"
+fi
+
```
I think it will be good to do this in the workflow, and keep this file agnostic of the CI. (Okay, it's currently called `cirrus.sh`, and we should anyway change this. ^^)

hebasto replied:
Thanks! [Resolved](https://github.com/bitcoin-core/secp256k1/pull/1396#issuecomment-1668502527).

real-or-random reviewed (commented):
Concept ACK.

Very nice that the prebuilt docker stuff can be moved to GitHub Actions easily. I think only time will tell whether we prefer GH Actions over Cirrus in terms of latency, reliability, support, etc, but I'm happy to try.

hebasto commented:
Updated 7ea529c40bc890cc8c398bb410413010350ee29b -> c066aff6cfedb3ad98cc8f76ad87da7a06a17011 ([pr1396.02](https://github.com/hebasto/secp256k1/commits/pr1396.02) -> [pr1396.03](https://github.com/hebasto/secp256k1/commits/pr1396.03), [diff](https://github.com/hebasto/secp256k1/compare/pr1396.02..pr1396.03)):

- addressed @MarcoFalke's [comment](https://github.com/bitcoin-core/secp256k1/pull/1396#discussion_r1285572943)
- addressed @real-or-random's [comment](https://github.com/bitcoin-core/secp256k1/pull/1396#discussion_r1286159462)

hebasto commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,26 @@
+      - name: Get modified files
+        id: changes
+        env:
+          COMMIT_BEFORE: ${{ github.event_name != 'pull_request' && github.event.pull_request.base.sha || (github.event.before != '0000000000000000000000000000000000000000' && github.event.before || github.event.repository.default_branch) }}
```
Fixed a typo `!=` --> `==`.

> Can you explain what this achieves?

By default, the [`actions/checkout`](https://github.com/actions/checkout) uses `fetch-depth: 1`.

We need to fetch `COMMIT_BEFORE` to figure out the modified files.

real-or-random replied:
It seems there's a built-in feature for this https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onpushpull_requestpull_request_targetpathspaths-ignore

hebasto replied:
> It seems there's a built-in feature for this [docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onpushpull_requestpull_request_targetpathspaths-ignore](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#onpushpull_requestpull_request_targetpathspaths-ignore)

It was the place I started from. However, I didn't find a design with that feature that allows to have two execution paths depending on whether docker image is invalidated.

hebasto commented:
> > GitHub Container Registry is free for public images.
> 
> Ok, so that will also need write-access for the CI again.

Not the whole CI, but a single job: https://github.com/bitcoin-core/secp256k1/blob/c066aff6cfedb3ad98cc8f76ad87da7a06a17011/.github/workflows/ci.yml#L72-L77 and its the corresponding reusable workflow: https://github.com/bitcoin-core/secp256k1/blob/c066aff6cfedb3ad98cc8f76ad87da7a06a17011/.github/workflows/docker.yml#L12-L13

> If there is no storage limit, it would still be good to clean old tags, no? I just can't imagine that GH will run an unlimited, free, append-only storage forever.

I agree with you. Going to leave this issue for a follow up as a non-critical one.

hebasto commented:
>    * Pull requests always use the cache from the main development branch and always build from scratch on a new push, if they modify the docker image.

Unfortunately, [`jobs.<job_id>.container.image`](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idcontainerimage) supports only remote images, not built locally.

hebasto commented:
> You could implement it yourself with three lines? Something like: `podman pull $remote_image_from_main` (to init the cache) and then `podman build ...` and `podman run cirrus.sh`

Sure. That is approach I'm currently working on :)

real-or-random commented:
> (re [#1396 (comment)](https://github.com/bitcoin-core/secp256k1/pull/1396#issuecomment-1669154100))
> 
> Or maybe there could be two tags: `main` and `random_pull`, where both are downloaded to opportunistically seed the cache. This would double the storage and require write permission in pull requests again, but would save on compute on a sunny day (no malicious pull request or honest concurrent pull request that modifies the docker image).

I think what you suggested above is good enough. It's anyway not a bad to rebuild every time in PRs that modify the Docker image, just to get rid of all potential caching issues during development.

real-or-random commented:
I meant this:
> Pull requests always use the cache from the main development branch and always build from scratch on a new push, if they modify the docker image.

---------

> IIRC clang is built from scratch, so if there is a matrix of 90 tasks, clang will be built 90 times in a pull, unless there happens to be a cache laying around.

Okay, sure, that's a problem. 

Could we have just one tag per PR, but delete the image when the PR is closed or merged?

real-or-random commented:
I'm not sure if publishing to ghcr is the right approach. Couldn't we use GHA's cache? https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows Then we get cache eviction automatically, see https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#usage-limits-and-eviction-policy.

I can imagine two approaches:
 * Cache the built image (similar to what we do now).
 * Always "build" the Docker image, but provide the workflow that builds the image with a Docker build cache. (Then `docker build` will always run, but it will immediately succeed when it has a cache hit...)

hebasto commented:
> I'm not sure if publishing to ghcr is the right approach. Couldn't we use GHA's cache? [docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows](https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows) Then we get cache eviction automatically, see [docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#usage-limits-and-eviction-policy](https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows#usage-limits-and-eviction-policy).
> 
> I can imagine two approaches:
> 
>     * Cache the built image (similar to what we do now).
> 
>     * Always "build" the Docker image, but provide the workflow that builds the image with a Docker build cache. (Then `docker build` will always run, but it will immediately succeed when it has a cache hit...)

I have already tried such an approach.

> ... it will immediately succeed...

That's a bit exaggerated :)

hebasto commented:
> > UPD. I mean, GBs of cache take time to pack / upload / download / unpack...
> 
> How and why is uploading an image to cache different from uploading an image to ghcr, time wise?

You're right. There are no significant difference. My point was that there is no reason to expect a success of a cached job "immediately".

real-or-random commented:
> > I can imagine two approaches:
> > 
> > * Cache the built image (similar to what we do now).
> > * Always "build" the Docker image, but provide the workflow that builds the image with a Docker build cache. (Then `docker build` will always run, but it will immediately succeed when it has a cache hit...)
> 
> I have already tried such an approach.

Which one have you tried, and why was it not successful?

hebasto commented:
> > > I can imagine two approaches:
> > > 
> > > * Cache the built image (similar to what we do now).
> > > * Always "build" the Docker image, but provide the workflow that builds the image with a Docker build cache. (Then `docker build` will always run, but it will immediately succeed when it has a cache hit...)
> > 
> > 
> > I have already tried such an approach.
> 
> Which one have you tried, and why was it not successful?

The latter one. Its recent incarnation is here: https://github.com/hebasto/secp256k1/tree/230808-gha-linux.3

It fails due to "no space left on device" error. However, the workflow works with simpler `linux-debian.Dockerfile`.

real-or-random commented:
> It fails due to "no space left on device" error

I can imagine this is due to our use of `/tmp`.

hebasto commented:
> It may be helpful to try `df -h` inside docker to see what's going on.

That seems a bit tricky as everything happens within a third-party [`docker/build-push-action`](https://github.com/docker/build-push-action).

Maybe just split MSVC part into a separated image?

hebasto commented:
@MarcoFalke 
> You can clear space, see [google/oss-fuzz@b7f04d7](https://github.com/google/oss-fuzz/commit/b7f04d782277638a67bc44865de445977eed4708)

Thank you! I'm going to look into your suggestion closely.

> Maybe just split MSVC part into a separated image?

Meanwhile, splitting Docker images has its own benefits like an opportunity to build all images in parallel and minimizing the total time needed to rebuild.

real-or-random commented:
There's also https://github.com/insightsengineering/disk-space-reclaimer (not suggesting to use that action, but this has a some nice stats etc.)

And yep, https://github.com/hebasto/secp256k1/commit/9db9c1ccd531d020d709af3e8b466b3fb250094c is also a good idea. We could also remove the gcc sources/build dir after installing.

hebasto commented:
Converting to draft.

Please consider https://github.com/bitcoin-core/secp256k1/pull/1398 first that uses GHA cache rather ghcr.io.

hebasto commented:
Rebased on top of the #1398.

real-or-random commented:
I think, it would be good to cherry-pick https://github.com/real-or-random/secp256k1/commit/20b208df0ab22a2afd80f3cf172f100365766e7b here to see the difference in image size.

hebasto commented:
Rebased on top of the #1399.

> I think, it would be good to cherry-pick [real-or-random@20b208d](https://github.com/real-or-random/secp256k1/commit/20b208df0ab22a2afd80f3cf172f100365766e7b) here to see the difference in image size.

Thanks! Cherry-picked.

hebasto commented:
FWIW, @real-or-random's commit reduced the cached Docker layers size from ~3.2 GB to ~2.1 GB.

hebasto commented:
Here is an example of the last commit's help: https://github.com/hebasto/secp256k1/actions/runs/5904122537/job/16015555795#step:3:621

real-or-random commented:
- Hm, yeah, sad that we need retries...(nit: Call I it `retry_builder`? `fallback` sounds like it's doing something else)
 - I wonder if artifacts will be a more reliable way of transferring the image from the builder job to the other job. Do you think it's worth trying? It would be interesting to see if it's slower or faster.

hebasto commented:
> * Hm, yeah, sad that we need retries...(nit: Call I it `retry_builder`? `fallback` sounds like it's doing something else)

Done.

>   * I wonder if artifacts will be a more reliable way of transferring the image from the builder job to the other job. Do you think it's worth trying? It would be interesting to see if it's slower or faster.

I agree with you. It's definitely worth trying.

Considering the 1st Sept. deadline, it would be nice to postpone until all tasks are migrated from Cirrus CI.

real-or-random commented:
> > *
> 
> I agree with you. It's definitely worth trying.
> 
> Considering the 1st Sept. deadline, it would be nice to postpone until all tasks are migrated from Cirrus CI.

Agreed, I'll add an item to the tracking issue.

real-or-random reviewed (approved):
ACK 4ad4914bd15bd856eddb306d86588bdacabb1184 (assuming CI finishes successfully)

real-or-random commented:
Well, now one job failed even with the retrying... Perhaps failure probability within a single job are correlated. https://github.com/bitcoin-core/secp256k1/actions/runs/5904705284/job/16017365422?pr=1396

Idk, I think we should accept some flakiness and move on for now... Cirrus was also not perfect in that regard, even though they improved their automatic re-running. So my ACK stands.

hebasto commented:
> I wonder if artifacts will be a more reliable way of transferring the image from the builder job to the other job.

Uploading of the image (2036732288 bytes) takes appr. 15 minutes

hebasto commented:
I took the liberty to push one more commit which partially reverts the change [introduced](https://github.com/bitcoin-core/secp256k1/pull/1398#issuecomment-1675888347) earlier and based on [this](https://github.com/bitcoin-core/secp256k1/pull/1398#issuecomment-1674495293) comment.

Actually, https://github.com/moby/buildkit/issues/3969 is related to writing to the GitHub Actions cache, but our `run-in-docker-action` only reads from it.

Before pushing, I've successfully run it 10 times in my personal repo -- https://github.com/hebasto/secp256k1/actions/runs/5912556017.

Here is an [example](https://github.com/hebasto/secp256k1/actions/runs/5912556017/job/16042929345) of how network glitches are currently being handled:
```
...
#11 sha256:6ee24e002191e89c757a065afeab642edbcab93b1d8c700e01985f03d349b115 598.74MB / 1.43GB 348.3s
#11 sha256:6ee24e002191e89c757a065afeab642edbcab93b1d8c700e01985f03d349b115 602.93MB / 1.43GB 353.4s
#11 353.8 error: failed to copy: read tcp 172.17.0.2:53852->20.209.52.129:443: read: connection reset by peer
#11 353.8 retrying in 1s
#11 sha256:6ee24e002191e89c757a065afeab642edbcab93b1d8c700e01985f03d349b115 604.54MB / 1.43GB 1.7s
#11 sha256:6ee24e002191e89c757a065afeab642edbcab93b1d8c700e01985f03d349b115 608.73MB / 1.43GB 6.9s
#11 sha256:6ee24e002191e89c757a065afeab642edbcab93b1d8c700e01985f03d349b115 613.97MB / 1.43GB 12.0s
...
```

hebasto commented:
Continuation of this PR is here: https://github.com/bitcoin-core/secp256k1/pull/1406.

real-or-random commented:
> Here is an [example](https://github.com/hebasto/secp256k1/actions/runs/5912556017/job/16042929345) of how network glitches are currently being handled:

Oh, nice! Just to get this right: the retries are only possible with the dropped `driver-opts`?

real-or-random commented:
Hm, 3 jobs failed during the Checkout stage with HTTP 429 (too many requests). I wonder if this is related to having so many parallel jobs. Have you seen this before, e.g. when testing in your repo?

hebasto commented:
> Hm, 3 jobs failed during the Checkout stage with HTTP 429 (too many requests). I wonder if this is related to having so many parallel jobs. Have you seen this before, e.g. when testing in your repo?

I've seen the same error in my personal repo exactly at the same time. I think, it was one of intermittent GitHub-wide glitches. FWIW, similar issues were observed in Cirrus CI a few times (still they were caused by GitHub).

real-or-random commented:
> > I wonder if artifacts will be a more reliable way of transferring the image from the builder job to the other job.
> 
> Uploading of the image (2036732288 bytes) takes appr. 15 minutes.
> 
> UPD. See builds in my personal repo -- [hebasto/secp256k1/actions/runs/5912496333](https://github.com/hebasto/secp256k1/actions/runs/5912496333).
> 
> UPD. Just a reminder that there is one more option that is "abusing" of GitHub Container Registry.

Thanks for testing the artifact thing. Downloading the image is much faster apparently (~ 1.3 min), and then the actual jobs run a bit faster. So I can imagine that approach is not too bad if we could avoid re-uploading existing artifacts, e.g., have another digest file that just stores the SHA256, and re-upload only if the SHA does not match. But all of this is probably not worth the complexity if the current approach with the cache turns out to be good enough.

real-or-random reviewed (approved):
ACK e10878f58e4022dbac6e215a89c980a17b95044b
